src/common/dom/apply_themes_on_element.ts

```diff
@@ -23,50 +23,85 @@ let PROCESSED_THEMES: Record<string, ProcessedTheme> = {};
  * Apply a theme to an element by setting the CSS variables on it.
  *
  * element: Element to apply theme on.
- * themes: HASS Theme information
- * selectedTheme: selected theme.
+ * themes: HASS theme information.
+ * selectedTheme: Selected theme.
+ * themeSettings: Settings such as selected dark mode and colors.
  */
 export const applyThemesOnElement = (
   element,
   themes: HomeAssistant["themes"],
   selectedTheme?: string,
-  themeOptions?: Partial<HomeAssistant["selectedTheme"]>
+  themeSettings?: Partial<HomeAssistant["selectedThemeSettings"]>
 ) => {
   let cacheKey = selectedTheme;
-  let themeRules: Partial<Theme> = {};
+  let themeRules: Partial<ThemeVars> = {};
 
-  if (selectedTheme === "default" && themeOptions) {
-    if (themeOptions.dark) {
+  if (themeSettings) {
+    // Determine the primary and accent colors. Fallbacks are the HA default blue and orange
+    // or the derived "darkStyles" values, depending on the light vs dark mode.
+    let primaryColor;
+    let accentColor;
+    if (selectedTheme === "default") {
+      // User selected colors
+      primaryColor = themeSettings.primaryColor;
+      accentColor = themeSettings.accentColor;
+    } else if (selectedTheme && themes.themes[selectedTheme]) {
+      // Try in that order:
+      // 1. Theme value from the selected mode
+      // 2. Mode-independent theme value
+      // 3. HA default colors (implicit)
+      if (themeSettings.dark) {
+        primaryColor =
+          themes.themes[selectedTheme].modes?.dark?.["primary-color"];
+        accentColor =
+          themes.themes[selectedTheme].modes?.dark?.["accent-color"];
+      } else {
+        primaryColor =
+          themes.themes[selectedTheme].modes?.light?.["primary-color"];
+        accentColor =
+          themes.themes[selectedTheme].modes?.light?.["accent-color"];
+      }
+
+      // Mode-independent theme value as fallback
+      primaryColor =
+        primaryColor || themes.themes[selectedTheme]["primary-color"];
+      accentColor =
+        primaryColor || themes.themes[selectedTheme]["accent-color"];
+    }
+
+    if (themeSettings.dark) {
       cacheKey = `${cacheKey}__dark`;
       themeRules = darkStyles;
-      if (themeOptions.primaryColor) {
+
+      if (primaryColor) {
         themeRules["app-header-background-color"] = hexBlend(
```

Comment:
Before doing this we should check if the theme doesn't contain a `app-header-background-color` value that will overrule this anyway